crt0stack/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+            }),
+            AuxvEntry::EGid(v) => self.auxv.push(AuxvPair {
+                key: AT_EGID,
+                value: v as _,
+            }),
+            AuxvEntry::Platform(v) => {
+                self.platform.replace(v);
+            }
+            AuxvEntry::HWCap(v) => self.auxv.push(AuxvPair {
+                key: AT_HWCAP,
+                value: v as _,
+            }),
+            AuxvEntry::ClockTick(v) => self.auxv.push(AuxvPair {
+                key: AT_CLKTCK,
+                value: v as _,
+            }),
+            AuxvEntry::Secure(v) => self.auxv.push(AuxvPair {
+                key: AT_SECURE,
+                value: if v { 1 } else { 0 },
+            }),
+            AuxvEntry::BasePlatform(v) => {
+                self.base_platform.replace(v);
+            }
+            AuxvEntry::Random(v) => {
+                self.random.replace(v);
+            }
+            AuxvEntry::HWCap2(v) => self.auxv.push(AuxvPair {
+                key: AT_HWCAP2,
+                value: v as _,
+            }),
+            AuxvEntry::ExecFilename(v) => {
+                self.exec_fn.replace(v);
+            }
+        }
+    }
+
+    /// serialize to a mutable byte slice
+    ///
+    ///
+    pub fn serialize<'a>(self, slice: &mut [u8]) -> Result<usize, Error> {
+        let mut sp = Stack::new(slice);
+        let mut str_ptr = Vec::new();
+        let mut rand_ptr: Option<*const u8> = None;
+        let mut platform_ptr: Option<*const u8> = None;
+        let mut base_platform_ptr: Option<*const u8> = None;
+        let mut exec_fn_ptr: Option<*const u8> = None;
+
+        str_ptr.push(core::ptr::null());
+        for s in self.envp {
+            sp.push_l(0u8);
+            sp.push_slice(s.as_bytes());
+            str_ptr.push(sp.ptr());
+        }
+        str_ptr.push(core::ptr::null());
+        for s in self.argv.iter().rev() {
+            sp.push_l(0u8);
+            sp.push_slice(s.as_bytes());
+            str_ptr.push(sp.ptr());
+        }
+
+        match self.platform {
+            None => {}
+            Some(x) => {
+                sp.push_l(0u8);
+                sp.push_slice(x.as_bytes());
+                platform_ptr = Some(sp.ptr());
+            }
+        }
+
+        match self.base_platform {
+            None => {}
+            Some(x) => {
+                sp.push_l(0u8);
+                sp.push_slice(x.as_bytes());
+                base_platform_ptr = Some(sp.ptr());
+            }
+        }
+
+        match self.exec_fn {
+            None => {}
+            Some(x) => {
+                sp.push_l(0u8);
+                sp.push_slice(x.as_bytes());
+                exec_fn_ptr = Some(sp.ptr());
+            }
+        }
+
+        match self.random {
+            None => {}
+            Some(x) => {
+                sp.align_down(core::mem::size_of::<AuxvType>());
+                sp.push_slice(&x);
+                rand_ptr = Some(sp.ptr());
+            }
+        }
```

Comment:
If we're not going to panic or anything when these are `None`, I think these can be compacted down with the [`if let` mechanism](https://doc.rust-lang.org/rust-by-example/flow_control/if_let.html), like this:

```diff
diff --git a/crt0stack/src/lib.rs b/crt0stack/src/lib.rs
index 5fbd02b..5bf9d99 100644
--- a/crt0stack/src/lib.rs
+++ b/crt0stack/src/lib.rs
@@ -224,12 +224,9 @@ impl Crt0Stack {
         sp.push_l(0 as AuxvType);
         sp.push_l(AT_NULL);
 
-        match rand_ptr {
-            None => {}
-            Some(p) => {
+        if let Some(p) = rand_ptr {
                 sp.push_l(p as AuxvType);
                 sp.push_l(AT_RANDOM);
-            }
         }
 
         match platform_ptr {
```

Reply:
done